extension/goplugin/environment.go

```diff
@@ -65,258 +65,269 @@ type SchemaPrioritizedSchemaHandlers map[string]PrioritizedSchemaHandlers
+	traceID string
 
-	initFnRaw plugin.Symbol
-	initFns   []func(goext.IEnvironment) error
+	handlers       EventPrioritizedHandlers
+	schemaHandlers EventSchemaPrioritizedSchemaHandlers
 
-	traceID string
+	rawTypes map[string]reflect.Type
+	types    map[string]reflect.Type
 }
 
 // NewEnvironment create new gohan extension rawEnvironment based on context
-func NewEnvironment(name string, dataStore db.DB, ident middleware.IdentityService, sync sync.Sync) *Environment {
-	newEnvironment := &Environment{
+func NewEnvironment(name string, beforeStartHook func(env *Environment) error, afterStopHook func(env *Environment)) *Environment {
+	env := &Environment{
+		initFns:         map[string]func(goext.IEnvironment) error{},
+		beforeStartHook: beforeStartHook,
+		afterStopHook:   afterStopHook,
+
 		name:  name,
-		db:    dataStore,
-		ident: ident,
-		sync:  sync,
+		ident: nil,
+
+		rawTypes: make(map[string]reflect.Type),
+		types:    make(map[string]reflect.Type),
 	}
-	return newEnvironment
+	return env
+}
+
+// Handlers returns a copy of handlers - used for testing
+func (env *Environment) Handlers() EventPrioritizedHandlers {
+	return env.handlers
+}
+
+// SchemaHandlers returns a copy of handlers - used for testing
+func (env *Environment) SchemaHandlers() EventSchemaPrioritizedSchemaHandlers {
+	return env.schemaHandlers
+}
+
+// RawTypes returns raw types
+func (env *Environment) RawTypes() map[string]reflect.Type {
+	return env.rawTypes
+}
+
+// Types returns types
+func (env *Environment) Types() map[string]reflect.Type {
+	return env.types
 }
 
 // Core returns an implementation to Core interface
-func (thisEnvironment *Environment) Core() goext.ICore {
-	return thisEnvironment.extCore
+func (env *Environment) Core() goext.ICore {
+	return env.coreImpl
 }
 
 // Logger returns an implementation to Logger interface
-func (thisEnvironment *Environment) Logger() goext.ILogger {
-	return thisEnvironment.extLogger
+func (env *Environment) Logger() goext.ILogger {
+	return env.loggerImpl
 }
 
 // Schemas returns an implementation to Schemas interface
-func (thisEnvironment *Environment) Schemas() goext.ISchemas {
-	return thisEnvironment.extSchemas
+func (env *Environment) Schemas() goext.ISchemas {
+	return env.schemasImpl
 }
 
 // Sync returns an implementation to Sync interface
-func (thisEnvironment *Environment) Sync() goext.ISync {
-	return thisEnvironment.extSync
+func (env *Environment) Sync() goext.ISync {
+	return env.syncImpl
 }
 
 // Database returns an implementation to IDatabase interface
-func (thisEnvironment *Environment) Database() goext.IDatabase {
-	return thisEnvironment.extDatabase
+func (env *Environment) Database() goext.IDatabase {
+	return env.databaseImpl
+}
+
+// HTTP returns an implementation to IHTTP interface
+func (env *Environment) HTTP() goext.IHTTP {
+	return &HTTP{}
+}
+
+// Auth returns an implementation to IAuth interface
+func (env *Environment) Auth() goext.IAuth {
+	return &Auth{}
 }
```

Comment:
`HTTP` and `Auth` do not have any relations to env. I am not sure if env should have these functions here. (just a comment. There is no need to change if you think it's unnecessary).

Reply:
Currently, HTTP does not require anything from Env but that may change at some moment in future - for example it may require some configuration from config file or authorization credentials or similar. Generally, the assumption is that whatever a plugin needs from gohan it should be available in Env. That gives a clear isolation by interfaces and in fact, guarantees that plugin compilation work correctly. I'd rather leave this as is.

Reply:
got it. it seems fine.